compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1512,9 +1514,118 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                            Some(hir::Node::Local(hir::Local { init: Some(init), .. })) => init,
+                            _ => return,
+                        };
+
+                        match pat_kind {
+                            hir::PatKind::Tuple(pats, ..) => match init.kind {
+                                ExprKind::Tup(init_tup) => {
+                                    for (pat, init) in zip(pats.iter(), init_tup.iter()) {
+                                        if pat.hir_id == *hir_id {
+                                            return self.note_type_is_not_clone(
+                                                diag,
+                                                expected_ty,
+                                                found_ty,
+                                                init,
+                                            );
+                                        }
+                                    }
+                                }
+                                ExprKind::Block(
+                                    hir::Block {
+                                        expr:
+                                            Some(Expr {
+                                                kind: ExprKind::Tup(init_block_expr_tup), ..
+                                            }),
+                                        ..
+                                    },
+                                    _,
+                                ) => {
+                                    for (pat, init) in zip(pats.iter(), init_block_expr_tup.iter())
+                                    {
+                                        if pat.hir_id == *hir_id {
+                                            return self.note_type_is_not_clone(
+                                                diag,
+                                                expected_ty,
+                                                found_ty,
+                                                init,
+                                            );
+                                        }
+                                    }
+                                }
+                                _ => {}
+                            },
+                            _ => {}
+                        }
+                    }
+                    _ => {}
+                };
+            }
+            // If we're a closure recurse back into that closure
+            hir::ExprKind::Closure(hir::Closure { body: body_id, .. }) => {
+                let hir::Body { value: body_expr, .. } = self.tcx.hir().body(*body_id);
+                return self.note_type_is_not_clone(diag, expected_ty, found_ty, body_expr);
+            }
+            // Similarly recurse into a block
+            hir::ExprKind::Block(hir::Block { expr: Some(block_expr), .. }, _) => {
+                return self.note_type_is_not_clone(diag, expected_ty, found_ty, block_expr);
+            }
```

Comment:
with the `trailing_expr` abstraction suggested above these duplication wouldn't be necessary i think.